### DISCUSSION TITLE
Expilitly identify each element of a server

### DISCUSSION THREAD

**Discussion by phillipp-schlueter** [FreeOpcUa/python-opcua#1550] [Q&A] [answered]
Is there a way to identify each element of a server explicitly? Display names and node ids are repetitive. I'm just wondering if I need to implement this myself or if there is something like an uid or specific attribute each element carries.

**Comment by schroeder-** [comment]
NodeIds are unique. But in the browse tree it possible that a node is referenced(linked) by multiple other nodes.

**Reply by phillipp-schlueter** [reply]
so `client.get_node()` could result in getting the same id twice? If so, is there a way to differentiate those?

**Reply by schroeder-** [reply]
No. NodeId is unique, so get_node returns only one Node. But in the Browsetree a node could be found multiple times.

**Reply by AndreasHeine** [reply, marked as answer]
it depends on the used hierarchical reference...
a node can be referenced multiple times so it will appear multiple times!

![image009](https://github.com/user-attachments/assets/dcb2c11d-d2e2-4d36-af70-1781b4dae069)

X:Sensor -> 3:SignalSet has a child X:Temperature (ProcessValueType)
X:Temperature is also referenced under X:Monitoring!

so if you crawl the X:MyMachine it will find it twice as long as you dont apply filter to the browse where you can specify the references you want the BrowseResult for!

**Reply by phillipp-schlueter** [reply]
But if I go over all elements of the server with a recursive function I get the same nodeid output twice.
```
def get_server_recursive(client, parent):
    children = client.get_node(parent).get_children()
    if children:
        for child in children:
            child_id = client.get_node(child)
            print(child_id)
            get_server_tree_recursive(client, child)
```
EDIT:
Sorry @AndreasHeine , I didn't see your reply befor commenting.
That makes sense, thanks!

**Reply by AndreasHeine** [reply]
get_children -> makes a browse request! https://github.com/FreeOpcUa/python-opcua/blob/master/opcua/common/node.py#L294-L316

`get_children(self, refs=ua.ObjectIds.HierarchicalReferences, nodeclassmask=ua.NodeClass.Unspecified):`

which calls:
`
    def get_referenced_nodes(self, refs=ua.ObjectIds.References, direction=ua.BrowseDirection.Both, nodeclassmask=ua.NodeClass.Unspecified, includesubtypes=True):`

includesubtypes=True means all subtypes of "ua.ObjectIds.HierarchicalReferences" 

ua.NodeClass:
https://github.com/FreeOpcUa/python-opcua/blob/master/opcua/ua/uaprotocol_auto.py#L356-L385

```
    Unspecified = 0
    Object = 1
    Variable = 2
    Method = 4
    ObjectType = 8
    VariableType = 16
    ReferenceType = 32
    DataType = 64
    View = 128
```
